docs/graph-homomorphism-as-natural-transformation.md

```diff
@@ -58,6 +68,10 @@ We then equate this to a graph by stating that:
  * A graph consists of vertexes V.
  * A graph consists of edges E, which are defined as products of vertexes V x V.
 
+[Paul: Note: the edge set E is a subset of V x V, not necessarily all of V x V.  Indeed, the edge set is all of V x V if and only if G is the complete directed graph on the vertex set V.
+
+Actually, if one wants to consider directed multigraphs (multiple edges between the same pair of vertices are allowed) then you can't really even think of E as a subset of V x V.  Instead you think of E as just an abstract set equipped with a map E -> V x V which sends every edge e to its source and target.  This fits well with the functorial definition of a graph.]
```

Comment:
It is normal to think of an edge in a graph as fundamentally bound to its source and target vertices, and a "label" on an edge as a bit of extra structure.  But in my opinion it is better to think of an edge as fundamentally a "label", and the source and target vertices are extra structure applied to it.  Interestingly, in CS this is consistent with how some graph libraries are implemented: there is a class `EdgeType` which stores some data (and which you can inherit from if you want to store more data), and you insert an `EdgeType` object into a graph by binding it to two vertices.

So with this perspective, I would look at a multigraph as follows.  You start with a vertex set V = {v1, v2, ..., vn} and an edge set E = {e1, e2, ..., en} - at the moment these are just two abstract sets which have nothing much to do with each other.  To build a graph out of these sets, you just have to define a map g: E -> V x V; using this map you can now think of an edge e as a connection from the first vertex in g(e) to the second.  And now there is no problem contemplating multiple edges between the same pair of vertices: even if g(e1) = g(e2), e1 and e2 retain their identities as distinct elements of E.

This is an appealing way to look at it for the categorically-minded as well: fundamentally, the graph "is" a morphism E -> V x V.  If you wanted you could consider a "graph object" in a monoidal category other than **Set** - it would simply be a pair of objects X, Y together with a morphism X -> Y x Y.